src/material/datepicker/calendar.ts

```diff
@@ -428,7 +417,16 @@ export class MatCalendar<D> implements AfterContentInit, AfterViewChecked, OnDes
   }
 
   /** Returns the component instance that corresponds to the current calendar view. */
-  private _getCurrentViewComponent() {
-    return this.monthView || this.yearView || this.multiYearView;
+  private _getCurrentViewComponent(): MatMonthView<D> | MatYearView<D> | MatMultiYearView<D> {
+    // The return type must be a union to ensure that the Closure compiler does not optimize the
+    // calls to _init(). See https://github.com/angular/components/issues/22996.
+    switch (this.currentView) {
```

Comment:
Is this `switch` required for the Closure compiler fix? The comment sounds like only the return type change would be necessary.

Reply:
Only the return type change is necessary.

However, `switch` makes it clear that we chose the current view component and does not rely on implicit falsyness of class fields that are non-nullable. The code had two ways of selecting the component (`if`s and `||`), I unified them.

The comment is to ensure that future refactorings do not revert to `||` and reduce the return type to only one type. IMO, it is an easy mistake to remove the return type of:
```
returnValue(): A | B {
  return this.a || this.b;
}
```

Reply:
It might just be me (also since I wrote the code to begin with), but I think I prefer the `||` since it's more compact and it doesn't depend on matching class instances to strings.

Reply:
Ok, changed back to use `||` and extended the comment with little more context.

I assume the existing code that did the string to view mapping (L391-L397) does not need to be restored, so I keep them unified.